sdk/metric/internal/filter.go

```diff
@@ -62,3 +76,44 @@ func (f *filter[N]) Aggregate(measurement N, attr attribute.Set) {
 func (f *filter[N]) Aggregation() metricdata.Aggregation {
 	return f.aggregator.Aggregation()
 }
+
+// precomputedFilter is an aggregator that applies attribute filter when
+// Aggregating for precomputed Aggregations. The precomputed Aggregations need
```

Comment:
This comment confuses me slightly.  Why does it matter whether filtering is done?  No filter is equivalent to an empty filter, and just means a single element is aggregated.  I would think that a precomputed Aggregation would need to _always sum_, but would only end up summing once when no filtering is done.  The sum needs to be reset every aggregation "cycle".  Is this an optimization to prevent resetting and storing each time?

Reply:
Precomputed aggregators don't add values, they set the measured value directly. The user is reporting the sum to record. They say the sun should be X and then in the next cycle they say the sun should be Y. If they added it would effectively saying in the second collection cycle the sum isn't Y but X+Y.

The problem is when one of these pre-computed sums has attributes removed and the resulting attributes collide with another pre-computed sum. In this case, the pre-computed some that had attributes removed needs to be added to the other.

Reply:
To illustrate my point, this branch doesn't seem to impact the resulting behavior: https://github.com/open-telemetry/opentelemetry-go/commit/0fecb363a260565719f401667846e7a490725467

It still passes the complete tests in meter_test.go

Reply:
The distinction between no filter and a noop filter means adding a filter that does nothing changes the resulting metrics, which seems wrong: https://github.com/open-telemetry/opentelemetry-go/commit/48dadf5e52f8a27917c48dc537cca6b8ef7d4c60

Reply:
> The distinction between no filter and a noop filter means adding a filter that does nothing changes the resulting metrics, which seems wrong: [48dadf5](https://github.com/open-telemetry/opentelemetry-go/commit/48dadf5e52f8a27917c48dc537cca6b8ef7d4c60)

I was not expecting that error in the comments, I expected the behavior to not change when introducing the no-op filter as well. I'm not able to reproduce the error though.

I pushed af451873e7aa903b86428e3d9b2b126ada4f572a to duplicate this in the upstream CI system to see if my issue with not producing the error is local.

Reply:
> To illustrate my point, this branch doesn't seem to impact the resulting behavior: [0fecb36](https://github.com/open-telemetry/opentelemetry-go/commit/0fecb363a260565719f401667846e7a490725467)
> 
> It still passes the complete tests in meter_test.go

I'm also not able to duplicate this change without the tests in `meter_test.go` failing.

```terminal 
--- FAIL: TestAsynchronousExample (0.00s)
    --- FAIL: TestAsynchronousExample/Cumulative (0.00s)
        meter_test.go:1207: [ScopeMetrics Metrics not equal:
            missing expected values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:60}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:53}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:5}}, Temporality:0x1, IsMonotonic:true}}
            unexpected additional values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:0}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:53}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 480373935, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 480735399, time.Local), Value:5}}, Temporality:0x1, IsMonotonic:true}}
            ]
    --- FAIL: TestAsynchronousExample/Delta (0.00s)
        meter_test.go:1282: [ScopeMetrics Metrics not equal:
            missing expected values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:0}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:6}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Time:time.Date(1, time.January, 1, 0, 0, 0, 0, time.UTC), Value:5}}, Temporality:0x2, IsMonotonic:true}}
            unexpected additional values:
            metricdata.Metrics{Name:"pageFaults", Description:"", Unit:"", Data:metricdata.Sum[int64]{DataPoints:[]metricdata.DataPoint[int64]{metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x1, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:-60}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x2, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:6}, metricdata.DataPoint[int64]{Attributes:attribute.Set{equivalent:attribute.Distinct{iface:[2]attribute.KeyValue{attribute.KeyValue{Key:"pid", Value:attribute.Value{vtype:4, numeric:0x0, stringly:"1001", slice:interface {}(nil)}}, attribute.KeyValue{Key:"tid", Value:attribute.Value{vtype:2, numeric:0x3, stringly:"", slice:interface {}(nil)}}}}}, StartTime:time.Date(2023, time.January, 17, 9, 45, 20, 481064632, time.Local), Time:time.Date(2023, time.January, 17, 9, 45, 20, 481101372, time.Local), Value:5}}, Temporality:0x2, IsMonotonic:true}}
            ]
FAIL
FAIL	go.opentelemetry.io/otel/sdk/metric	0.025s
```

This is expected as a duplicate `Observer` of the same attribute set that is not filtered should be overwritten, not added (i.e. #3278 and #3350).

Reply:
> I pushed [af45187](https://github.com/open-telemetry/opentelemetry-go/commit/af451873e7aa903b86428e3d9b2b126ada4f572a) to duplicate this in the upstream CI system to see if my issue with not producing the error is local.

These changes passed upstream as well: https://github.com/open-telemetry/opentelemetry-go/actions/runs/3941799369/jobs/6744664098

Reply:
I'm wondering if our disconnect is in the base behavior here. If there is a `Int64ObservableCounter` for an SDK that filters its attributes to only allow the `user` attribute and the following observations are made:

- `obserer.Observe(10, {"user", "alice"})`
- `obserer.Observe(30, {"user", "alice"})`

The resulting measurement expected `30` for the sum datapoint with the `{"user", "alice"}` attributes, right?

Reply:
I hadn't considered that you could record the same set of labels twice within an observation.  I'm used to that being disallowed in prom :).  But it would break this case.

Reply:
I was wrong above as well, sorry.  I was running with both patches, rather than just the test commit.

Reply:
Ok, so if I have the observations below, I get value = 30.
```
obserer.Observe(10)
obserer.Observe(30)
```

But then if I have the observations below, and a view that filters `key != foo`, I would get value = 40?

```
obserer.Observe(10, {"foo", "bar"})
obserer.Observe(30, {"foo", "bar"})
```

Fundamentally I want a filtered-out attribute to be the same as an attribute that never existed.  That implies that the fact that something was or was not filtered should never matter (in my ideal world).  Only the post-filter attribute set should matter.

Reply:
> Fundamentally I want a filtered-out attribute to be the same as an attribute that never existed. That implies that the fact that something was or was not filtered should never matter (in my ideal world). Only the post-filter attribute set should matter. So in theory, I would probably rather have the first example sum within a collection cycle.

So, to be clear, I agree with the approach you mention. However, this is not the behavior the broader OTel community expects as captured in https://github.com/open-telemetry/opentelemetry-specification/issues/2905 (and previously in https://github.com/open-telemetry/opentelemetry-specification/issues/1874).

The specification expects behavior laid out in [the supplemental guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/25f75bfde209630c60728039f27c041f70e5e738/specification/metrics/supplementary-guidelines.md#asynchronous-example-attribute-removal-in-a-view). Specifically, _additive_ for any precomputed sum that has had attributes removed.

I'm happy to switch the behavior if upstream the specification decides to do so, but with this PR I'm only trying to move our implementation into compliance with what the specification guidelines are.

Reply:
https://github.com/open-telemetry/opentelemetry-specification/issues/3109

Reply:
lets resolve this conversation and move forward with the fix (assuming other approvers are on-board)